File: Sample/Engine/Gameplay/TileMap.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import logging
import os
from typing import Any, Dict, List, Optional
from .. import (
    Vector2i,
    Vector2u,
    Color,
    Image,
    Texture,
    ReturnType,
    TileLayerGraphics,
    AutoTile,
    TileLayerData,
    Shader,
)
from ..Utils.Inner import IS_IOS_PLATFORM, warnIosShaderSkippedOnce


class TileLayer(TileLayerGraphics):
    r"""Tile layer renderable.

    Renders a tile grid using a `VertexArray` and provides
    per-cell queries for passability and material properties.
    Autotile cells are rendered with the standard RPG Maker XP 48-state
    composition through additional vertex arrays managed in the C++
    backend.
    """

    def __init__(
        self,
        data: TileLayerData,
        texture: Texture,
        autoTileTextures: Optional[List[Texture]] = None,
        autoTileFrameCounts: Optional[List[int]] = None,
        visible: bool = True,
    ) -> None:
        r"""Initialise a TileLayer.

        - \param data                TileLayerData containing the tile-index grid and autotile pool
        - \param texture             Texture (tileset image) for rendering
        - \param autoTileTextures    Per-autotile-pool textures matching `data.autoTilePool`
        - \param autoTileFrameCounts Per-autotile-pool animation frame counts
        - \param visible             Whether the layer is initially visible
        """
        from .. import CellSize

        self._data = data
        self._width = len(self._data.tiles[0]) if self._data.tiles else 0
        self._height = len(self._data.tiles)
        self._texture = texture
        self._lightBlockMapCache: Optional[List[List[float]]] = None
        self._lightBlockImageCache: Optional[Image] = None
        self._reflectionStrengthMapCache: Optional[List[List[float]]] = None
        self._reflectionStrengthImageCache: Optional[Image] = None
        self._ignoreLightingMapCache: Optional[List[List[float]]] = None
        self._ignoreLightingImageCache: Optional[Image] = None
        self.visible = visible
        self.shaderPath: str = str(self._data.shaderPath or "")
        self.shader: Optional[Shader] = None
        self.shaderType: Optional[Shader.Type] = None
        self._shaderTime: float = 0.0
        self._shaderUsesTime: bool = False
        self._loadShader()

        autoTilePool = list(self._data.autoTilePool or [])
        if not self._data.autoTileKeys:
            self._data.autoTileKeys = [entry.name for entry in autoTilePool]
        autoTileTextureList: List[Texture] = list(autoTileTextures or [])
        if len(autoTileTextureList) < len(autoTilePool):
            autoTileTextureList.extend([None] * (len(autoTilePool) - len(autoTileTextureList)))
        autoTileMaterials = [(entry.material if entry is not None else None) for entry in autoTilePool]
        autoTileFrames = list(autoTileFrameCounts or [])
        if len(autoTileFrames) < len(autoTilePool):
            autoTileFrames.extend([1] * (len(autoTilePool) - len(autoTileFrames)))
        self._autoTileTextures = autoTileTextureList
        self._autoTileFrameCounts = autoTileFrames
        self._autoTileMaterialsRef = autoTileMaterials

        super().__init__(
            self._width,
            self._height,
            CellSize,
            self._texture,
            self._data,
            autoTileTextureList,
            autoTileFrames,
        )

    def _loadShader(self) -> None:
        if not self.shaderPath:
            return
        if IS_IOS_PLATFORM:
            warnIosShaderSkippedOnce(
                f"TileLayer.shaderPath:{self.shaderPath}",
                f"iOS: shaders are disabled; skipped tile layer shader: {self.shaderPath}",
            )
            return
        extension = os.path.splitext(self.shaderPath)[1].lower()
        shaderTypes = {
            ".vert": Shader.Type.Vertex,
            ".frag": Shader.Type.Fragment,
        }
        shaderType = shaderTypes.get(extension)
        if shaderType is None:
            logging.warning("Unsupported tile layer shader extension: %s", self.shaderPath)
            return
        fullPath = self.shaderPath
        normalizedPath = fullPath.replace("\\", "/").lstrip("./")
        if not normalizedPath.lower().startswith("assets/shaders/"):
            fullPath = os.path.join(".", "Assets", "Shaders", self.shaderPath)
        if not os.path.exists(fullPath):
            logging.warning("Tile layer shader file not found: %s", fullPath)
            return
        try:
            with open(fullPath, "r", encoding="utf-8") as shaderFile:
                self._shaderUsesTime = "uniform float time" in shaderFile.read()
            self.shader = Shader(fullPath, shaderType)
            self.shaderType = shaderType
        except OSError as exc:
            logging.warning("Tile layer shader file could not be read for %s: %s", self.shaderPath, exc)
        except Exception as exc:
            logging.warning("Tile layer shader load failed for %s: %s", self.shaderPath, exc)

    def updateShader(self, deltaTime: float) -> None:
        r"""Advance the conventional time uniform for this layer shader.

        - \param deltaTime Elapsed time in seconds
        """
        if self.shader is None or not self._shaderUsesTime:
            return
        self._shaderTime += deltaTime
        self.shader.setUniform("time", self._shaderTime)

    @ReturnType(name=str)
    def getName(self) -> str:
        r"""Return the layer name.

        - \return  The `layerName` string
        """
        return self._data.layerName

    @ReturnType(tiles=List[List[Optional[int]]])
    def getTiles(self) -> List[List[Optional[int]]]:
        r"""Return the tile-index grid.

        - \return  2D list of tile indices (`None` = empty cell)
        """
        return self._data.tiles

    @ReturnType(autoTiles=List[List[Optional[int | str]]])
    def getAutoTiles(self) -> List[List[Optional[int | str]]]:
        r"""Return the autotile pool-index grid.

        - \return  2D list of autotile pool indices (`None` = no autotile)
        """
        return self._data.autoTiles

    @ReturnType(pool=List[AutoTile])
    def getAutoTilePool(self) -> List[AutoTile]:
        r"""Return the autotile pool used by this layer.

        - \return  List of `AutoTile` entries referenced by `autoTiles`
        """
        return self._data.autoTilePool

    @ReturnType(lightBlock=float)
    def getLightBlock(self, position: Vector2i) -> float:
        r"""Return the light-block value at the given position.

        - \param position  Grid position to query
        - \return          Light-block factor (0.0–1.0)
        """
        return self.getMaterialProperty(position, "lightBlock")

    @ReturnType(mirror=bool)
    def getMirror(self, position: Vector2i) -> bool:
        r"""Check whether the tile mirrors at the given position.

        - \param position  Grid position to query
        - \return          `True` if the surface mirrors
        """
        return self.getMaterialProperty(position, "mirror")

    @ReturnType(reflectionStrength=float)
    def getReflectionStrength(self, position: Vector2i) -> float:
        r"""Return the reflection strength at the given position.

        - \param position  Grid position to query
        - \return          Reflection strength (0.0–1.0)
        """
        return self.getMaterialProperty(position, "reflectionStrength")

    @ReturnType(ignoreLighting=bool)
    def getIgnoreLighting(self, position: Vector2i) -> bool:
        r"""Check whether the tile ignores ambient and direct lighting.

        - \param position  Grid position to query
        - \return          `True` if the surface is rendered unlit
        """
        return bool(self.getMaterialProperty(position, "ignoreLighting"))

    @ReturnType(speedRate=Optional[float])
    def getSpeedRate(self, position: Vector2i) -> Optional[float]:
        r"""Return the movement speed rate at the given position.

        - \param position  Grid position to query
        - \return          Speed multiplier, or `None`
        """
        return self.getMaterialProperty(position, "speedRate")

    def getMaterialProperty(self, position: Vector2i, propertyName: str) -> Any:
        r"""Return an arbitrary material property at the given position.

        - \param position      Grid position to query
        - \param propertyName  Name of the material attribute to read
        - \return              Property value, or `None` if no material
        """
        material = self.getMaterial(position)
        if material is None:
            return None
        return getattr(material, propertyName, None)

    def getLightBlockMap(self) -> List[List[float]]:
        r"""Build or return the cached light-block map.

        - \return  2D grid of light-block values (0.0–1.0)
        """
        if self._lightBlockMapCache is None:
            self._lightBlockMapCache = super().getLightBlockMap()
        return self._lightBlockMapCache

    def getReflectionStrengthMap(self) -> List[List[float]]:
        r"""Build or return the cached reflection-strength map.

        - \return  2D grid of reflection strength values
        """
        if self._reflectionStrengthMapCache is None:
            self._reflectionStrengthMapCache = super().getReflectionStrengthMap()
        return self._reflectionStrengthMapCache

    def getIgnoreLightingMap(self) -> List[List[float]]:
        r"""Build or return the cached ignore-lighting map.

        - \return  2D grid of ignore-lighting values
        """
        if self._ignoreLightingMapCache is None:
            self._ignoreLightingMapCache = super().getIgnoreLightingMap()
        return self._ignoreLightingMapCache

    def getLightBlockImage(self) -> Image:
        r"""Build or return the cached light-block image.

        - \return  Grayscale `Image` where brightness = light block
        """
        if self._lightBlockImageCache is None:
            dataMap = self.getLightBlockMap()
            img = Image(Vector2u(self._width, self._height))
            for y in range(self._height):
                for x in range(self._width):
                    g = int(dataMap[y][x] * 255)
                    img.setPixel(Vector2u(x, y), Color(g, g, g))
            self._lightBlockImageCache = img
        return self._lightBlockImageCache

    def getReflectionStrengthImage(self) -> Image:
        r"""Build or return the cached reflection-strength image.

        - \return  Grayscale `Image` where brightness = reflection strength
        """
        if self._reflectionStrengthImageCache is None:
            dataMap = self.getReflectionStrengthMap()
            img = Image(Vector2u(self._width, self._height))
            for y in range(self._height):
                for x in range(self._width):
                    g = int(dataMap[y][x] * 255)
                    img.setPixel(Vector2u(x, y), Color(g, g, g))
            self._reflectionStrengthImageCache = img
        return self._reflectionStrengthImageCache

    def getIgnoreLightingImage(self) -> Image:
        r"""Build or return the cached ignore-lighting image.

        - \return  Grayscale `Image` where brightness = ignore-lighting
        """
        if self._ignoreLightingImageCache is None:
            dataMap = self.getIgnoreLightingMap()
            img = Image(Vector2u(self._width, self._height))
            for y in range(self._height):
                for x in range(self._width):
                    g = int(dataMap[y][x] * 255)
                    img.setPixel(Vector2u(x, y), Color(g, g, g))
            self._ignoreLightingImageCache = img
        return self._ignoreLightingImageCache


class Tilemap:
    r"""Tilemap container.

    Holds multiple named `TileLayer`s and provides
    aggregated queries across all layers.
    """

    def __init__(self, layers: List[TileLayer]) -> None:
        r"""Initialise a Tilemap from a list of layers.

        - \param layers  List of `TileLayer` instances to register
        """
        self._layers: Dict[str, TileLayer] = {}
        for layer in layers:
            self._layers[layer.getName()] = layer
        self._tilesData: Dict[str, List[List[Optional[int]]]] = {}
        for layerName, layer in self._layers.items():
            self._tilesData[layerName] = layer.getTiles()

    @ReturnType(layer=Optional[TileLayer])
    def getLayer(self, name: str) -> Optional[TileLayer]:
        r"""Return a layer by name.

        - \param name   Name of the layer to retrieve
        - \return         The `TileLayer`, or `None`
        """
        for layerName, layer in self._layers.items():
            if layerName == name:
                return layer
        return None

    @ReturnType(tiles=Dict[str, List[List[Optional[int]]]])
    def getTilesData(self) -> Dict[str, List[List[Optional[int]]]]:
        r"""Return raw tile-index data for all layers.

        - \return  Dictionary mapping layer name to tile-index grid
        """
        return self._tilesData

    @ReturnType(autoTiles=Dict[str, List[List[Optional[int | str]]]])
    def getAutoTilesData(self) -> Dict[str, List[List[Optional[int | str]]]]:
        r"""Return raw autotile pool-index data for every layer.

        - \return  Dictionary mapping layer name to autotile pool-index grid
        """
        return {layerName: layer.getAutoTiles() for layerName, layer in self._layers.items()}

    @ReturnType(layers=Dict[str, TileLayer])
    def getAllLayers(self) -> Dict[str, TileLayer]:
        r"""Return all layers.

        - \return  Dictionary mapping layer name to `TileLayer`
        """
        return self._layers

    @ReturnType(layerNames=List[str])
    def getLayerNameList(self) -> List[str]:
        r"""Return all layer names.

        - \return  List of layer name strings
        """
        return list(self._layers.keys())

    @ReturnType(size=Vector2u)
    def getSize(self) -> Vector2u:
        r"""Return the map dimensions in tiles.

        - \return  `(width, height)` as `Vector2u`
        """
        if len(self._layers) == 0:
            return Vector2u(0, 0)
        first = next(iter(self._layers.values()))
        return Vector2u(first._width, first._height)

    def updateAutoTileAnimation(self, deltaTime: float, frameInterval: float = 0.5) -> None:
        r"""Advance autotile animation across every layer.

        - \param deltaTime      Elapsed time in seconds
        - \param frameInterval  Seconds between two animation frames
        """
        for layer in self._layers.values():
            layer.updateAutoTileAnimation(deltaTime, frameInterval)
            layer.updateShader(deltaTime)

File: Sample/Core/Engine/src/Gameplay/TileMap.cpp
#include <Gameplay/TileMap.hpp>

#include <Runtime/EngineState.hpp>
#include <Utils/Inner.hpp>
#include <Utils/ShaderLoader.hpp>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <stdexcept>
#include <utility>

TileLayer::TileLayer(
    const TileLayerData& data, std::shared_ptr<sf::Texture> texture,
    const std::vector<std::shared_ptr<sf::Texture>>& autoTileTextures,
    const std::vector<int>& autoTileFrameCounts, bool visible, bool deferred)
    : TileLayerGraphics(layerWidth(data), layerHeight(data), CellSize,
                        requireTexture(texture), data,
                        normalizeAutoTileTextures(data, autoTileTextures),
                        normalizeFrameCounts(data, autoTileFrameCounts),
                        deferred),
      visible(visible),
      shaderPath(data.shaderPath),
      data_(data),
      width_(layerWidth(data)),
      height_(layerHeight(data)),
      texture_(std::move(texture)),
      autoTileTextures_(normalizeAutoTileTextures(data, autoTileTextures)),
      autoTileFrameCounts_(normalizeFrameCounts(data, autoTileFrameCounts)) {
    loadShader();
}

TileLayer::~TileLayer() = default;

const std::string& TileLayer::getName() const {
    return data_.layerName;
}

const TileGrid& TileLayer::getTiles() const {
    return data_.tiles;
}

const AutoTileGrid& TileLayer::getAutoTiles() const {
    return data_.autoTiles;
}

const std::vector<AutoTile>& TileLayer::getAutoTilePool() const {
    return data_.autoTilePool;
}

std::optional<std::string> TileLayer::getAutoTileKey(int poolIndex) const {
    if (poolIndex < 0 ||
        poolIndex >= static_cast<int>(data_.autoTileKeys.size())) {
        return std::nullopt;
    }
    return data_.autoTileKeys[static_cast<std::size_t>(poolIndex)];
}

TileLayerData TileLayer::getData() const {
    return data_;
}

void TileLayer::writeBlock(int x, int y, const TileGrid& tileBlock,
                           const AutoTileGrid& autoTileBlock) {
    data_.validateBlock(x, y, tileBlock, autoTileBlock);
    writePendingBlock(x, y, tileBlock, autoTileBlock);
    data_.writeBlock(x, y, tileBlock, autoTileBlock);
    lightBlockMapCache_.reset();
    reflectionStrengthMapCache_.reset();
    ignoreLightingMapCache_.reset();
    lightBlockImageCache_.reset();
    reflectionStrengthImageCache_.reset();
    ignoreLightingImageCache_.reset();
}

std::vector<std::shared_ptr<sf::Texture>> TileLayer::getAutoTileTextures()
    const {
    return autoTileTextures_;
}

std::vector<int> TileLayer::getAutoTileFrameCounts() const {
    return autoTileFrameCounts_;
}

std::shared_ptr<TileLayer> TileLayer::rebuild(
    const TileLayerData& data,
    const std::vector<std::shared_ptr<sf::Texture>>& autoTileTextures,
    const std::vector<int>& autoTileFrameCounts) const {
    return std::make_shared<TileLayer>(data, texture_, autoTileTextures,
                                       autoTileFrameCounts, visible);
}

bool TileLayer::getVisible() const {
    return visible;
}

void TileLayer::setVisible(bool value) {
    visible = value;
}

bool TileLayer::hasContent(const sf::Vector2i& position) const {
    return get(position).has_value() || getAutoTileAt(position).has_value();
}

bool TileLayer::isDirectionPassable(const sf::Vector2i& position,
                                    int directionIndex) const {
    const std::optional<int> tile = get(position);
    if (!tile.has_value()) {
        return true;
    }
    if (directionIndex < 0 || directionIndex >= 4) {
        return false;
    }
    const std::size_t tileIndex = static_cast<std::size_t>(*tile);
    if (tileIndex >= data_.layerTileset.dir4.size()) {
        return true;
    }
    const auto& directions = data_.layerTileset.dir4[tileIndex];
    if (directions.size() != 4) {
        return true;
    }
    return directions[static_cast<std::size_t>(directionIndex)];
}

std::optional<MaterialValue> TileLayer::getMaterialProperty(
    const sf::Vector2i& position, const std::string& propertyName) const {
    const std::optional<Material> material = getMaterial(position);
    if (!material.has_value()) {
        return std::nullopt;
    }

    std::string fieldName = propertyName;
    if (propertyName == "getLightBlock") {
        fieldName = "lightBlock";
    } else if (propertyName == "getMirror") {
        fieldName = "mirror";
    } else if (propertyName == "getReflectionStrength") {
        fieldName = "reflectionStrength";
    } else if (propertyName == "getIgnoreLighting") {
        fieldName = "ignoreLighting";
    } else if (propertyName == "getSpeedRate") {
        fieldName = "speedRate";
    }

    const MaterialData values = material->asDict();
    const auto iterator = values.find(fieldName);
    return iterator == values.end()
               ? std::nullopt
               : std::optional<MaterialValue>(iterator->second);
}

std::optional<float> TileLayer::getLightBlock(
    const sf::Vector2i& position) const {
    return materialFloat(getMaterialProperty(position, "lightBlock"));
}

std::optional<bool> TileLayer::getMirror(const sf::Vector2i& position) const {
    return materialBool(getMaterialProperty(position, "mirror"));
}

std::optional<float> TileLayer::getReflectionStrength(
    const sf::Vector2i& position) const {
    return materialFloat(getMaterialProperty(position, "reflectionStrength"));
}

std::optional<bool> TileLayer::getIgnoreLighting(
    const sf::Vector2i& position) const {
    return materialBool(getMaterialProperty(position, "ignoreLighting"));
}

std::optional<float> TileLayer::getSpeedRate(
    const sf::Vector2i& position) const {
    return materialFloat(getMaterialProperty(position, "speedRate"));
}

void TileLayer::updateShader(float deltaTime) {
    if (shader == nullptr || !shaderUsesTime_) {
        return;
    }
    shaderTime_ += deltaTime;
}

void TileLayer::draw(sf::RenderTarget& target, sf::RenderStates states) const {
    if (shader != nullptr && shaderUsesTime_ && states.shader == shader.get()) {
        shader->setUniform("time", shaderTime_);
    }
    TileLayerGraphics::draw(target, states);
}

std::shared_ptr<sf::Shader> TileLayer::getShader() const {
    return shader;
}

std::shared_ptr<sf::Image> TileLayer::getLightBlockImage() {
    if (lightBlockImageCache_ == nullptr) {
        lightBlockImageCache_ = buildMaterialImage(getLightBlockMapView());
    }
    return lightBlockImageCache_;
}

std::shared_ptr<sf::Image> TileLayer::getReflectionStrengthImage() {
    if (reflectionStrengthImageCache_ == nullptr) {
        if (!reflectionStrengthMapCache_.has_value()) {
            reflectionStrengthMapCache_ =
                TileLayerGraphics::getReflectionStrengthMap();
        }
        reflectionStrengthImageCache_ =
            buildMaterialImage(*reflectionStrengthMapCache_);
    }
    return reflectionStrengthImageCache_;
}

std::shared_ptr<sf::Image> TileLayer::getIgnoreLightingImage() {
    if (ignoreLightingImageCache_ == nullptr) {
        if (!ignoreLightingMapCache_.has_value()) {
            ignoreLightingMapCache_ = TileLayerGraphics::getIgnoreLightingMap();
        }
        ignoreLightingImageCache_ =
            buildMaterialImage(*ignoreLightingMapCache_);
    }
    return ignoreLightingImageCache_;
}

sf::Vector2u TileLayer::getGridSize() const {
    return {static_cast<unsigned int>(std::max(0, width_)),
            static_cast<unsigned int>(std::max(0, height_))};
}

bool TileLayer::isCellBuilt(const sf::Vector2i& position) const {
    return TileLayerGraphics::isCellBuilt(position);
}

const std::vector<std::vector<float>>& TileLayer::getLightBlockMapView() {
    if (!lightBlockMapCache_.has_value()) {
        lightBlockMapCache_ = TileLayerGraphics::getLightBlockMap();
    }
    return *lightBlockMapCache_;
}

int TileLayer::layerWidth(const TileLayerData& data) {
    return data.tiles.empty() ? 0 : static_cast<int>(data.tiles.front().size());
}

int TileLayer::layerHeight(const TileLayerData& data) {
    return static_cast<int>(data.tiles.size());
}

const std::shared_ptr<sf::Texture>& TileLayer::requireTexture(
    const std::shared_ptr<sf::Texture>& texture) {
    if (texture == nullptr) {
        throw std::invalid_argument("TileLayer texture must not be null");
    }
    return texture;
}

std::vector<std::shared_ptr<sf::Texture>> TileLayer::normalizeAutoTileTextures(
    const TileLayerData& data,
    const std::vector<std::shared_ptr<sf::Texture>>& textures) {
    std::vector<std::shared_ptr<sf::Texture>> result = textures;
    result.resize(data.autoTilePool.size());
    return result;
}

std::vector<int> TileLayer::normalizeFrameCounts(
    const TileLayerData& data, const std::vector<int>& frameCounts) {
    std::vector<int> result = frameCounts;
    result.resize(data.autoTilePool.size(), 1);
    for (int& frameCount : result) {
        if (frameCount <= 0) {
            frameCount = 1;
        }
    }
    return result;
}

std::optional<float> TileLayer::materialFloat(
    const std::optional<MaterialValue>& value) {
    if (!value.has_value()) {
        return std::nullopt;
    }
    const float* number = std::get_if<float>(&*value);
    return number == nullptr ? std::nullopt : std::optional<float>(*number);
}

std::optional<bool> TileLayer::materialBool(
    const std::optional<MaterialValue>& value) {
    if (!value.has_value()) {
        return std::nullopt;
    }
    const bool* boolean = std::get_if<bool>(&*value);
    return boolean == nullptr ? std::nullopt : std::optional<bool>(*boolean);
}

void TileLayer::loadShader() {
    if (shaderPath.empty()) {
        return;
    }
    const std::optional<sf::Shader::Type> shaderType =
        ShaderLoader::inferType(shaderPath);
    if (!shaderType.has_value()) {
        std::cerr << "Unsupported tile layer shader extension: " << shaderPath
                  << '\n';
        return;
    }

    ShaderLoadResult result = ShaderLoader::load(shaderPath, *shaderType);
    if (!result) {
        std::cerr << result.error << '\n';
        return;
    }
    shader = std::move(result.shader);
    shaderUsesTime_ =
        result.source.find("uniform float time") != std::string::npos;
}

std::shared_ptr<sf::Image> TileLayer::buildMaterialImage(
    const std::vector<std::vector<float>>& values) const {
    std::shared_ptr<sf::Image> image =
        std::make_shared<sf::Image>(getGridSize());
    for (int y = 0; y < height_; ++y) {
        for (int x = 0; x < width_; ++x) {
            const float value = y < static_cast<int>(values.size()) &&
                                        x < static_cast<int>(values[y].size())
                                    ? values[y][x]
                                    : 0.0f;
            const int channel = std::clamp(
                static_cast<int>(std::floor(value * 255.0f)), 0, 255);
            image->setPixel(
                {static_cast<unsigned int>(x), static_cast<unsigned int>(y)},
                sf::Color(channel, channel, channel));
        }
    }
    return image;
}

Tilemap::Tilemap(const std::vector<std::shared_ptr<TileLayer>>& layers) {
    for (const std::shared_ptr<TileLayer>& layer : layers) {
        addLayer(layer);
    }
}

Tilemap::~Tilemap() = default;

void Tilemap::addLayer(const std::shared_ptr<TileLayer>& layer) {
    if (layer == nullptr) {
        throw std::invalid_argument("Tilemap layer must not be null");
    }
    const std::string name = layer->getName();
    if (!layers_.contains(name)) {
        layerNames_.push_back(name);
    }
    layers_.insert_or_assign(name, layer);
}

std::shared_ptr<TileLayer> Tilemap::getLayer(const std::string& name) const {
    const auto iterator = layers_.find(name);
    return iterator == layers_.end() ? nullptr : iterator->second;
}

std::unordered_map<std::string, TileGrid> Tilemap::getTilesData() const {
    std::unordered_map<std::string, TileGrid> result;
    for (const std::string& name : layerNames_) {
        const std::shared_ptr<TileLayer> layer = getLayer(name);
        if (layer != nullptr) {
            result.emplace(name, layer->getTiles());
        }
    }
    return result;
}

std::unordered_map<std::string, std::shared_ptr<TileLayer>>
Tilemap::getAllLayers() const {
    return layers_;
}

std::vector<std::string> Tilemap::getLayerNameList() const {
    return layerNames_;
}

std::unordered_map<std::string, AutoTileGrid> Tilemap::getAutoTilesData()
    const {
    std::unordered_map<std::string, AutoTileGrid> result;
    for (const std::string& name : layerNames_) {
        const std::shared_ptr<TileLayer> layer = getLayer(name);
        if (layer != nullptr) {
            result.emplace(name, layer->getAutoTiles());
        }
    }
    return result;
}

sf::Vector2u Tilemap::getSize() const {
    if (layerNames_.empty()) {
        return {0, 0};
    }
    const std::shared_ptr<TileLayer> layer = getLayer(layerNames_.front());
    return layer == nullptr ? sf::Vector2u(0, 0) : layer->getGridSize();
}

void Tilemap::updateAutoTileAnimation(float deltaTime, float frameInterval) {
    for (const std::string& name : layerNames_) {
        const std::shared_ptr<TileLayer> layer = getLayer(name);
        if (layer == nullptr) {
            continue;
        }
        layer->updateAutoTileAnimation(deltaTime, frameInterval);
        layer->updateShader(deltaTime);
    }
}
